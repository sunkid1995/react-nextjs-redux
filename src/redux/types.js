export const PREFIXES = {
  AUTH: 'auth/',
  SERVICE_API: 'serviceApi/',
  SHARED: 'shared/',
};

export const INIT_ACTION = { type: '@@redux/INIT' };

export const AUTH = {
  AUTHORIZE: 'auth/AUTHORIZE',
  DEAUTHORIZE: 'auth/DEAUTHORIZE',
};

export const SERVICE_API = {
  AUTH_WITH_EMAIL: 'serviceApi/AUTH_WITH_EMAIL',
  FORGOT_PASSWORD: 'serviceApi/FORGOT_PASSWORD',
  RESET_PASSWORD: 'serviceApi/RESET_PASSWORD',
};

export const SHARED = {

};
