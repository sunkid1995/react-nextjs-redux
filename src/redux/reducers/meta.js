import { REHYDRATE } from 'redux-persist/lib/constants';

const INITIAL_STATE = {
  rehydrated: false,
};

export default function metaReducer(state = INITIAL_STATE, action) {
  const { type } = action;

  if (type === REHYDRATE) {
    const { err } = action;

    /**
     * On server-side, REHYDRATE action will throw an err:
     * Error: No available storage method found
     * Otherwise on client-side, REHYDRATE will be dispatched without error
     * So we check if err is nullable then it's actually rehydrated in client
     */
    if (err == null) return { ...state, rehydrated: true };
  }

  return state;
}
