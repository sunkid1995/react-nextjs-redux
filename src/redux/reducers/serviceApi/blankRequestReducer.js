import _ from 'lodash';

// Constants
import { AXIOS_REQUEST_SUFFIXES } from 'src/constants';

const { ERROR, SUCCESS } = AXIOS_REQUEST_SUFFIXES;

export default function blankRequestReducer(state, action) {
  const { type } = action;

  if (_.endsWith(type, ERROR)) {
    const { error, meta: { previousAction } } = action;
    const { payload: { dataKey } } = previousAction;
    return { ...state, [dataKey]: { error, loading: false } };
  }

  if (_.endsWith(type, SUCCESS)) {
    const { meta: { previousAction } } = action;
    const { payload: { dataKey } } = previousAction;
    return { ...state, [dataKey]: { error: undefined, loading: false } };
  }

  const { payload: { dataKey } } = action;
  if (dataKey == null) return state;
  return { ...state, [dataKey]: { error: undefined, loading: true } };
}
