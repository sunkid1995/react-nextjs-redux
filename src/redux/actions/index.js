import * as auth from './auth';
import * as serviceApi from './serviceApi';
import * as shared from './shared';

export { auth, serviceApi, shared };
