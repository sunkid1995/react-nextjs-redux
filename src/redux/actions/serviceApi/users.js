import { createAction } from 'redux-actions';

// Constants
import { API_CONFIGS } from 'src/constants';

// Models
import User from 'src/models/User';

// Redux
import { SERVICE_API } from 'src/redux/types';

// Utils
// import { buildHeaders, responseInterceptor } from 'src/redux/utils';

const { REQUEST_METHODS: { PATCH, POST } } = API_CONFIGS;

export function authWithEmail({ email, password }) {
  const action = createAction(SERVICE_API.AUTH_WITH_EMAIL);
  const dataKey = 'emailAuth';
  const manifest = { klass: User.klass };

  return dispatch => {
    const request = {
      data: { email, password },
      method: POST,
      url: '/users/login',
    };

    return dispatch(action({ dataKey, manifest, request }));
  };
}

export function forgotPassword(email) {
  const action = createAction(SERVICE_API.FORGOT_PASSWORD);
  const dataKey = 'passwordForgotten';
  const manifest = { reserved: true };

  return dispatch => {
    const request = {
      data: { email },
      method: POST,
      url: '/users/password',
    };

    return dispatch(action({ dataKey, manifest, request }));
  };
}

export function resetPassword({ password, passwordConfirm, token }) {
  const action = createAction(SERVICE_API.RESET_PASSWORD);
  const dataKey = 'passwordReset';
  const manifest = { reserved: true };

  return dispatch => {
    const data = {
      password,
      reset_password_token: token,
      password_confirmation: passwordConfirm,
    };

    const request = {
      data,
      method: PATCH,
      url: '/users/password',
    };

    return dispatch(action({ dataKey, manifest, request }));
  };
}

