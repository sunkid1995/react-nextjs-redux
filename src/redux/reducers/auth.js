// Redux
import { AUTH } from 'src/redux/types';

const INITIAL_STATE = {
  user: undefined,
};

export default function authReducer(state = INITIAL_STATE, action) {
  const { type } = action;

  if (type === AUTH.AUTHORIZE) {
    const { payload: user } = action;
    return { ...state, user };
  }

  if (type === AUTH.DEAUTHORIZE) {
    return INITIAL_STATE;
  }

  return state;
}
