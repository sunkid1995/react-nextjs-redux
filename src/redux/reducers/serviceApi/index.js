import _ from 'lodash';

// Constants
import { AUTH, SERVICE_API } from 'src/redux/types';

// Locals
// import blankRequestReducer from './blankRequestReducer';
// import paginatedRequestReducer from './paginatedRequestReducer';
import plainRequestReducer from './plainRequestReducer';

const INITIAL_BLANK_REQUEST = { error: undefined, loading: false };
// const INITIAL_PAGINATED_DATA_REQUEST = { dataPages: {}, totalPages: 1, ...INITIAL_BLANK_REQUEST };
const INITIAL_PLAIN_DATA_REQUEST = { data: {}, ...INITIAL_BLANK_REQUEST };

const INITIAL_STATE = {
  emailAuth: INITIAL_PLAIN_DATA_REQUEST,
  passwordForgotten: INITIAL_PLAIN_DATA_REQUEST,
  passwordReset: INITIAL_PLAIN_DATA_REQUEST,
};

export default function serviceApiReducer(state = INITIAL_STATE, action) {
  const { type } = action;

  if (type === AUTH.DEAUTHORIZE) return INITIAL_STATE;

  if (
    _.startsWith(type, SERVICE_API.AUTH_WITH_EMAIL) ||
    _.startsWith(type, SERVICE_API.FORGOT_PASSWORD) ||
    _.startsWith(type, SERVICE_API.RESET_PASSWORD)
  ) return plainRequestReducer(state, action);

  return state;
}
