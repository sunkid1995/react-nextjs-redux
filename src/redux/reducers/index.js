import auth from './auth';
import meta from './meta';
import serviceApi from './serviceApi';
import shared from './shared';

export { auth, meta, serviceApi, shared };
