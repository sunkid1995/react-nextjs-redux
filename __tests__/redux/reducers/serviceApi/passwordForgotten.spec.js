import { serviceApi } from 'src/redux/reducers';
import { SERVICE_API } from 'src/redux/types';

// Helpers
import buildMockState from '__tests__/helpers/buildMockState';
import { plainDataRequest } from '__tests__/helpers/buildMockState/serviceApi';
import { buildErrorAction, buildSuccessAction } from '__tests__/helpers/mockApiRequests';

describe('redux/reducers/serviceApi/passwordForgotten', () => {
  const { serviceApi: mockState } = buildMockState();
  const mockData = {};

  const requestAction = { type: SERVICE_API.FORGOT_PASSWORD, payload: { dataKey: 'passwordForgotten' } };
  const successAction = buildSuccessAction(requestAction, mockData);
  const { errorAction } = buildErrorAction(requestAction, { noResponse: true });

  test('start', () => {
    const expectedState = {
      ...mockState,
      passwordForgotten: { ...plainDataRequest, loading: true },
    };

    const reduced = serviceApi(mockState, requestAction);
    expect(reduced).toEqual(expectedState);
    expect(reduced).toMatchSnapshot();
  });

  test('success', () => {
    const expectedState = {
      ...mockState,
      passwordForgotten: { ...plainDataRequest, data: mockData },
    };

    const reduced = serviceApi(mockState, successAction);
    expect(reduced).toEqual(expectedState);
    expect(reduced).toMatchSnapshot();
  });

  test('error', () => {
    const expectedState = {
      ...mockState,
      passwordForgotten: {
        ...plainDataRequest,
        error: { message: `${SERVICE_API.FORGOT_PASSWORD} request failed` },
      },
    };

    const reduced = serviceApi(mockState, errorAction);
    expect(reduced).toEqual(expectedState);
    expect(reduced).toMatchSnapshot();
  });

  test('refresh', () => {
    const currentState = {
      ...mockState,
      passwordForgotten: { ...plainDataRequest, data: mockData },
    };

    const expectedState = {
      ...mockState,
      passwordForgotten: { ...plainDataRequest, data: mockData, loading: true },
    };

    const reduced = serviceApi(currentState, requestAction);
    expect(reduced).toEqual(expectedState);
    expect(reduced).toMatchSnapshot();
  });
});
