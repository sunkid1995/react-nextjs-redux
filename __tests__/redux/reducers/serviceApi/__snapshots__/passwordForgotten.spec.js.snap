// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux/reducers/serviceApi/passwordForgotten error 1`] = `
Object {
  "emailAuth": Object {
    "data": undefined,
    "error": undefined,
    "loading": false,
  },
  "passwordForgotten": Object {
    "data": undefined,
    "error": Object {
      "message": "serviceApi/FORGOT_PASSWORD request failed",
      "response": undefined,
    },
    "loading": false,
  },
  "passwordReset": Object {
    "data": undefined,
    "error": undefined,
    "loading": false,
  },
}
`;

exports[`redux/reducers/serviceApi/passwordForgotten refresh 1`] = `
Object {
  "emailAuth": Object {
    "data": undefined,
    "error": undefined,
    "loading": false,
  },
  "passwordForgotten": Object {
    "data": Object {},
    "error": undefined,
    "loading": true,
  },
  "passwordReset": Object {
    "data": undefined,
    "error": undefined,
    "loading": false,
  },
}
`;

exports[`redux/reducers/serviceApi/passwordForgotten start 1`] = `
Object {
  "emailAuth": Object {
    "data": undefined,
    "error": undefined,
    "loading": false,
  },
  "passwordForgotten": Object {
    "data": undefined,
    "error": undefined,
    "loading": true,
  },
  "passwordReset": Object {
    "data": undefined,
    "error": undefined,
    "loading": false,
  },
}
`;

exports[`redux/reducers/serviceApi/passwordForgotten success 1`] = `
Object {
  "emailAuth": Object {
    "data": undefined,
    "error": undefined,
    "loading": false,
  },
  "passwordForgotten": Object {
    "data": Object {},
    "error": undefined,
    "loading": false,
  },
  "passwordReset": Object {
    "data": undefined,
    "error": undefined,
    "loading": false,
  },
}
`;
